### DISCUSSION TITLE
Returning a promise from before_dispatch results in Inactivity timeout

### DISCUSSION THREAD

**Discussion by aimass** [mojolicious/mojo#2156] [Q&A] [answered]
I'm writing an authorization filter using Mojolicious::Plugin::OAuth2 and a **before_dispatch** hook.
The hook validates the Access and/or OIDC tokens and make decisions if to allow the request or not (using multiple IdP). 
All works great so far... Then I started implementing an auto-renew feature so if the Renewal Token is still good, we can renew the token from within the hook.
Mojolicious::Plugin::OAuth2::get_refresh_token_p uses promises to renew the tokens. I thought that simply by using $c->render_later and returning the promise from the hook, would continue the request when the promise is fulfilled. 
But instead it just times out at the end of the promise; not sure how to get the request continue.
I also tried around_dispatch and $next->() but that didn't work either (I put the $next->() where the comment  " ###WHAT CAN I CALL HERE..." is)

In the hook:

```
    TOKEN_EXPIRED:
    if($self->auto_renew){
        $c->render_later;
        return $self->_auto_renew($c);
    }
```

And this is the auto-renew sub (see comment "### WHAT CAN I CALL HERE..." ):
```
sub _auto_renew ($self, $c){
    my $provider = $c->session('selected_provider');
    my $data = {refresh_token => $c->session('refresh_token')};
    return $c->oauth2->get_refresh_token_p($provider => $data)->then(sub {
        my $provider_res = shift;

        # Renew failed
        unless ($provider_res){
            $c->session(expires => 1);
            $self->_bail_out($c);
            return;
        }

        # Token received
        $c->session(access_token => $provider_res->{access_token});
        $c->session(expiration => $ENV{SESSION_EXPIRY});
        $c->session(id_token => $provider_res->{id_token});
        $c->session(scope => $provider_res->{scope});
        $c->session(token_type => $provider_res->{token_type});
        $c->session(refresh_token => $provider_res->{refresh_token});

        ### WHAT CAN I CALL HERE TO CONTINUE INSTEAD OF TIMEOUT ???
        
    })->catch(sub {
        my $error = shift;
        $c->log->debug("get_refresh_token_p returned error: $error");
        $self->bail_out($c);
        return undef;
    });

}
```

Any help or pointers greatly appreciated ! 

Best,

-- 
Alex

**Comment by rabbiveesh** [comment]
You aren't rendering anything, so that's why it times out.

have you tried something like  `$c->render(json => { msg => 'ok!' })`?

**Reply by aimass** [reply]
Yeah, but this is a before_dispatch hook. Normally you want to return without rendering anything (unless exception), and allow the router continue. If I don't use $c->render_later, the request continues before the token is renewed and it doesn't make it to the cookie on time.

**Reply by aimass** [reply]
This worked:
```
$c->app->routes->dispatch($c);
```
But this didn't:
```
$c->app->routes->continue($c);
```
Do you have any idea why ??

**Comment by aimass** [comment, marked as answer]
This worked:
```
$c->app->routes->dispatch($c);
```
I put this where the "### WHAT CAN I CALL HERE..." comment is and it works in my case.

But this did not work:
```
$c->app->routes->continue($c)
```
Any ideas why? I would think that continue() would be the right call in a filter but maybe this is more like starting a new request. Does anybody know exactly why continue() doesn't work in this case?
